docs_nnx/guides/linen_to_nnx.rst

```diff
@@ -0,0 +1,732 @@
+
+* NNX modules are stateful and automatically tracks a few things within, such as RNG keys and BatchNorm stats. That's why you don't need to explicitly pass an RNG key in on every step. Note that you can use `nnx.reseed <https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/rnglib.html#flax.nnx.reseed>`__ to reset its underlying RNG state.
+
+* In Linen, you need to explicitly define and pass in an argument ``training`` to control the behavior of ``nn.Dropout`` (namely, its ``deterministic`` flag, which means random dropout only happens if ``training=True``). In NNX, you can call ``model.train()`` to automatically switch ``nnx.Dropout`` to training mode. Conversely, call ``model.eval()`` to turn off training mode. You can learn more about what this API does at its `API reference <https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/module.html#flax.nnx.Module.train>`__.
+
+
+.. codediff::
+  :title: Linen, NNX
+  :sync:
+
+  ...
+
+  @jax.jit
+  def train_step(key, params, inputs, labels):
+    def loss_fn(params):
+      logits = model.apply(
+        {'params': params},
+        inputs, training=True, # <== inputs
+        rngs={'dropout': key}
+      )
+      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+
+    grads = jax.grad(loss_fn)(params)
+
+    params = jax.tree.map(lambda p, g: p - 0.1 * g, params, grads)
+    return params
+
+  ---
+
+  model.train() # Sets ``deterministic=False` under the hood for nnx.Dropout
+
+  @nnx.jit
+  def train_step(model, inputs, labels):
+    def loss_fn(model):
+      logits = model(
+
+        inputs, # <== inputs
+
+      )
+      return optax.softmax_cross_entropy_with_integer_labels(logits, labels).mean()
+
+    grads = nnx.grad(loss_fn)(model)
+    _, params = nnx.split(model)
+    params = jax.tree.map(lambda p, g: p - 0.1 * g, params, grads)
+    nnx.update(model, params)
+
+.. testcode:: Linen
+  :hide:
+
+  train_step(jax.random.key(0), params, sample_x, jnp.ones((1,), dtype=jnp.int32))
+
+.. testcode:: NNX
+  :hide:
+
+  sample_x = jnp.ones((1, 784))
+  train_step(model, sample_x, jnp.ones((1,), dtype=jnp.int32))
+
+
+Collections and Variable Types
+==========
+
+One key difference between Linen and NNX APIs is how we group variables into categories. In Linen, we use different collections; in NNX, since all variables shall be top-level Python attributes, you use different variable types.
+
+You can freely create your own variable types as subclasses of ``nnx.Variable``.
+
+For all the built-in Flax Linen layers and collections, NNX already created the corresponding layers and variable type. For example:
+
+ * ``nn.BatchNorm`` creates ``batch_stats`` -> ``nnx.BatchNorm`` creates ``nnx.BatchStats``.
+
+ * ``linen.Module.sow()`` creates ``intermediates`` -> ``nnx.Module.sow()`` creates ``nnx.Intermediates``. You can also simply get the intermediates by assigning it to a module attribute, like ``self.sowed = nnx.Intermediates(x)``.
```

Comment:
```suggestion
 * ``nn.Dense`` creates ``params`` -> ``nnx.Linear`` creates ``nnx.Param``.
  * ``nn.BatchNorm`` creates ``batch_stats`` -> ``nnx.BatchNorm`` creates ``nnx.BatchStat``.
 * ``linen.Module.sow()`` creates ``intermediates`` -> ``nnx.Module.sow()`` creates ``nnx.Intermediates``. You can also simply get the intermediates by assigning it to a module attribute, like ``self.sowed = nnx.Intermediate(x)``.
```
BTW: in Linen this is also true, you can simply add an intermediate via `self.variable('intermediates' 'sowed', lambda: x)`